File: pipeline/elicit.py
"""Elicit API client — literature (and clinical-trial) retrieval.

Elicit searches 138M+ academic papers and ClinicalTrials.gov. It does NOT cover
patents (verified against docs.elicit.com, 2026-07); patents must come from a
different source (AMASS patentcore, PatentsView, Lens, ...).

Auth: Bearer key (elk_live_...), Pro plan or above. Docs: https://docs.elicit.com/
"""
from __future__ import annotations

import os

import requests
from dotenv import load_dotenv

BASE = "https://elicit.com/api/v1"


class ElicitError(RuntimeError):
    pass


def _key() -> str:
    load_dotenv()
    k = os.environ.get("ELICIT_API_KEY")
    if not k:
        raise ElicitError("No ELICIT_API_KEY in env (add it to .env).")
    return k


def _post(path: str, body: dict) -> dict:
    r = requests.post(
        f"{BASE}{path}",
        json=body,
        headers={"Authorization": f"Bearer {_key()}", "Content-Type": "application/json"},
        timeout=90,
    )
    if r.status_code == 401:
        raise ElicitError("401 from Elicit: check ELICIT_API_KEY / plan tier.")
    r.raise_for_status()
    return r.json()


def search_papers(query: str, max_results: int = 10, corpus: str = "elicit",
                  search_mode: str = "semantic", filters: dict | None = None) -> dict:
    """POST /search. corpus: 'elicit' (138M) or 'pubmed'. Returns raw JSON.

    NOTE: the response wrapper key must be confirmed against a live call before
    downstream code depends on it (do not assume the top-level shape). Per-paper
    fields are: title, authors, year, abstract, doi, pmid, elicitId, venue,
    citedByCount, urls.
    """
    body: dict = {"query": query, "maxResults": max_results, "corpus": corpus,
                  "searchMode": search_mode}
    if filters:
        body["filters"] = filters
    return _post("/search", body)


def search_trials(query: str, max_results: int = 10) -> dict:
    """POST /search/trials. Returns raw JSON (shape TBD, confirm on live call)."""
    return _post("/search/trials", {"query": query, "maxResults": max_results})


if __name__ == "__main__":
    import json
    import sys

    q = sys.argv[1] if len(sys.argv) > 1 else "GLP-1 receptor agonist obesity clinical outcomes"
    print(json.dumps(search_papers(q, max_results=3), indent=2)[:3000])

File: web/src/lib/elicit.ts
// Elicit literature search (server-side). Papers + clinical trials only, not
// patents. Bearer auth. Docs: https://docs.elicit.com/
const BASE = "https://elicit.com/api/v1";

export interface ElicitPaper {
  title: string;
  authors: string[];
  year: number | null;
  abstract: string | null;
  doi: string | null;
  venue: string | null;
  citedByCount: number | null;
  urls: string[];
}

export async function searchPapers(query: string, maxResults = 8): Promise<ElicitPaper[]> {
  const key = process.env.ELICIT_API_KEY;
  if (!key) throw new Error("Missing ELICIT_API_KEY");

  const res = await fetch(`${BASE}/search`, {
    method: "POST",
    headers: { Authorization: `Bearer ${key}`, "Content-Type": "application/json" },
    body: JSON.stringify({ query, maxResults, corpus: "elicit", searchMode: "semantic" }),
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Elicit ${res.status}: ${(await res.text()).slice(0, 200)}`);

  const data = await res.json();
  return (data.papers ?? []).map(
    (p: Record<string, unknown>): ElicitPaper => ({
      title: String(p.title ?? ""),
      authors: (p.authors as string[]) ?? [],
      year: (p.year as number) ?? null,
      abstract: (p.abstract as string) ?? null,
      doi: (p.doi as string) ?? null,
      venue: (p.venue as string) ?? null,
      citedByCount: (p.citedByCount as number) ?? null,
      urls: (p.urls as string[]) ?? [],
    })
  );
}
